Promote recommendation to its own module

piuccio opened the pull request:
## What does this change?

* refactor: move the recommendation logic in its own file and report errors
* feature: show a tooltip when anonymous users click on recommend

## What is the value of this and can you measure success?

Maybe more logins? Actually it's just because a button doing nothing is stupid.

## Does this affect other platforms - Amp, Apps, etc?

Nope

## Screenshots

![screen shot 2016-08-01 at 17 32 23](https://cloud.githubusercontent.com/assets/680284/17301551/7ff0b29c-580f-11e6-8b0a-134c9702a29f.png)


## Request for comment

@NathanielBennett I've reuse the CSS of `report` comment, moving it to a generic `tooltip` module.

@sndrs not sure if I can use atomic CSS already

The wording might require some work, UX is on it.

## TODO

 - [x] Unit tests
 - [x] UX

NathanielBennett commented:
:+1: Code looks good. But I thought we were going to allow anonymous recommends. Which kind of changes the use case. @crifmulholland what you saying here?

regiskuckaertz commented on discussion/app/views/fragments/recommendationTooltip.scala.html:
```diff
@@ -0,0 +1,16 @@
+<div class="tooltip-box tooltip-box--neutral tooltip-box--width4 tooltip-box-hidden js-rec-tooltip">
```
Generally speaking, every DOM fragment inside a page that is hidden should have either the `hidden` or `aria-hidden="true"` attributes. All these `*-hidden` classes are only for people with sight 😭

regiskuckaertz commented on static/src/javascripts/projects/common/modules/discussion/upvote.js:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    function isOpenForRecommendations (element) {
+        return !!element.querySelector('.d-discussion--recommendations-open');
```
You could use `getElementsByClassName` which is much faster, but I don't know if there is a polyfill for IE8

regiskuckaertz commented:
Nice! Q: why is the upvote link even there when the user is not logged in?

piuccio commented:
Great question. The upvote link gives a context. I've tried removing the button but then it was hard to understand that `123` referred to the number of upvotes.

The arrow is not clickable but visually it still looks like a button, so it deserves some feedback when you click it.

NathanielBennett commented on discussion/app/views/fragments/reportComment.scala.html:
```diff
@@ -1,7 +1,7 @@
+    <div hidden class="d-discussion__error js-discussion__report-comment-error">
         <i class="i i-alert"></i>
-        <span class="d-discussion__error-text">Sorry there was an error. Please try again later. If the problem persists, please <a data-link-name="report comment eeeor : mailto : userhelp" href="mailto:userhelp@@theguardian.com">Userhelp</a></span>
+        <span class="d-discussion__error-text">Sorry there was an error. Please try again later. If the problem persists, please contact <a data-link-name="report comment error : mailto : userhelp" href="mailto:userhelp@@theguardian.com">Userhelp</a></span>
```
Thanks for correcting my spelling for me. :)

regiskuckaertz commented:
Fair enough. I would hide the number of votes too 😝

piuccio commented:
@crifmulholland I've added tracking on `sign in` and  `register` and modified the `data-link-name` of the button when user is not logged in.

@NathanielBennett I've pushed some changes, it's not using `--hidden` classes anymore but the `hidden` attribute.

I've added some unit tests and UX is ok with the wording, let's get this out

NathanielBennett commented:
@piuccio I don't know enough CS to doubt you. Looks good. 
Lets ship
:+1:
